### PR TITLE
Fixes #8100

### LINKED ISSUE #8100: [Regression] times.format interferes with strformat
Introduced in 371f87e9626ce84a55b3d8e1aa1bc3a97ebcc91b
```nim
import times, strformat

type Foo = object
    t: Time

var v: Foo
echo &"{v.t}"
```
```
test.nim(7, 6) template/generic instantiation from here
Projects/nim/lib/pure/strformat.nim(262, 11) Error: expression 'format(v.t, fmtRes238034, local())' is of type 'string' and has to be discarded
```

### PR DESCRIPTION
**Pull request by GULPF**
I looked for a less hacky way to check if a proc returns void, but couldn't find any. Related: #7397

### CODE REVIEW

**Comment by hlaaftana** [comment]
You could technically do `not compiles(type(format(arg, res)))` if that feature doesn't get implemented. Also you can do `compiles do: ` instead of `compiles((block: `

File: lib/pure/strformat.nim [hunk 1]
```diff
@@ -258,7 +258,9 @@ template callFormat(res, arg) {.dirty.} =
     # workaround in order to circumvent 'strutils.format' which matches
     # too but doesn't adhere to our protocol.
     res.add arg
-  elif compiles(format(arg, res)):
+  elif compiles(format(arg, res)) and
```
Comment (dom96) [inline lib/pure/strformat.nim]: Maybe the following would be better?

```nim
elif compiles((var res = ""; format(arg, res)))
```

Strange that the current code doesn't catch the "return value needs to be discarded" error. Maybe the above version will?

**Comment by GULPF** [comment]
> Strange that the current code doesn't catch the "return value needs to be discarded" error. Maybe the above version will?

It's because `compiles` expects an expression and not a statement, and expressions doesn't need to be discarded.

I switched to the `do` version because it looks slightly better IMO.

**Comment by Araq** [comment]
Travis hickup, closing and re-opening.

**Pull request merged**